src/runs/execution/tools/wikipedia-tool.ts

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright 2024 IBM Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { z } from 'zod';
+import { AnyTool } from 'bee-agent-framework/tools/base';
+import { WikipediaTool } from 'bee-agent-framework/tools/search/wikipedia';
+import { SimilarityTool } from 'bee-agent-framework/tools/similarity';
+import { splitString } from 'bee-agent-framework/internals/helpers/string';
+import * as R from 'remeda';
+import { cosineSimilarityMatrix } from 'bee-agent-framework/internals/helpers/math';
+
+import { defaultAIProvider } from '@/runs/execution/provider';
+
+const MAX_CONTENT_LENGTH_CHARS = 25_000;
+
+// Estimate from OpenAI: https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them
+const CHARS_PER_TOKEN_AVG = 4;
+
+// Creates a wikipedia tool that supports information retrieval
+export function wikipediaTool(
+  passageSizeTokens = 400,
+  overlapTokens = 100,
+  maxResults = 5
+): AnyTool {
+  // LLM to perform text embedding
+  const embeddingLLM = defaultAIProvider.createEmbeddingModel();
+
+  // Similarity tool to calculate the similarity between a query and a set of wikipedia passages
+  const similarity = new SimilarityTool({
+    maxResults,
+    provider: async (input): Promise<{ score: number }[]> => {
+      const embeds = await embeddingLLM.embed([
+        input.query,
+        ...input.documents.map((doc) => doc.text)
+      ]);
+      const similarities = cosineSimilarityMatrix(
+        [embeds.embeddings[0]], // Query
+        embeds.embeddings.slice(1) // Documents
+      )[0];
+      if (!similarities) {
+        throw new Error('Missing similarities');
+      }
+      return similarities.map((score) => ({ score }));
+    }
+  });
+
+  const wikipedia = new WikipediaTool({
+    filters: { minPageNameSimilarity: 0.25, excludeOthersOnExactMatch: false },
+    output: { maxSerializedLength: MAX_CONTENT_LENGTH_CHARS }
```

Comment:
`maxSerializedLength` is not affecting the markdown output.

I removed the restrictions on max content length from markdown as well as the simplified extraction:
```
extraction: { fields: { markdown: {} } },
```
Previously we had table extraction disabled and the output was truncated to 25k characters due to slow embeddings, however the issue has mostly been resolved and we can include this data again.

Also this makes our implementation more aligned with framework defaults.

Reply:
The `maxSerializedLength` property affects only the serialized output, which contains the markdown output and it works correctly.

```ts
    const instance = new WikipediaTool({
      output: {
        maxSerializedLength: 100,
      },
    });
    const response = await instance.run({
      query: "ice hockey",
    });
    expect(response.getTextContent()).toHaveLength(100);
```

Source: https://github.com/i-am-bee/bee-agent-framework/blob/6332b0c3a6cb82310fdcef7f576c4a5a6e2d40fd/src/tools/search/wikipedia.ts#L100-L111